bqskit/passes/processing/treescan.py

```diff
@@ -0,0 +1,226 @@
+                successful removal of a gate.
+                (Default: HilbertSchmidtResidualsGenerator())
+
+            instantiate_options (dict[str: Any]): Options passed directly
+                to circuit.instantiate when instantiating circuit
+                templates. (Default: {})
+
+            tree_depth (int): The depth of the tree of potential
+                solutions to instantiate. Note that 2^(tree_depth) - 1
+                circuits will be instantiated in parallel. Note that the default
+                behavior will be equivalent to normal ScanningGateRemoval
+                (Default: 1)
+
+            collection_filter (Callable[[Operation], bool] | None):
+                A predicate that determines which operations should be
+                attempted to be removed. Called with each operation
+                in the circuit. If this returns true, this pass will
+                attempt to remove that operation. Defaults to all
+                operations.
+        """
+
+        if not is_real_number(success_threshold):
+            raise TypeError(
+                'Expected real number for success_threshold'
+                ', got %s' % type(success_threshold),
+            )
+
+        if not isinstance(cost, CostFunctionGenerator):
+            raise TypeError(
+                'Expected cost to be a CostFunctionGenerator, got %s'
+                % type(cost),
+            )
+
+        if not isinstance(instantiate_options, dict):
+            raise TypeError(
+                'Expected dictionary for instantiate_options, got %s.'
+                % type(instantiate_options),
+            )
+
+        self.collection_filter = collection_filter or default_collection_filter
+
+        if not callable(self.collection_filter):
+            raise TypeError(
+                'Expected callable method that maps Operations to booleans for'
+                ' collection_filter, got %s.' % type(self.collection_filter),
+            )
+
+        if not is_integer(tree_depth):
+            raise TypeError(
+                'Expected Integer type for tree_depth, got %s.'
+                % type(instantiate_options),
+            )
+
+        self.tree_depth = tree_depth
+        self.start_from_left = start_from_left
+        self.success_threshold = success_threshold
+        self.cost = cost
+        self.instantiate_options: dict[str, Any] = {
+            'dist_tol': self.success_threshold,
+            'min_iters': 10,
+            'cost_fn_gen': self.cost,
+        }
+        self.instantiate_options.update(instantiate_options)
+
+    @staticmethod
+    def get_tree_circs(
+        orig_num_cycles: int,
+        circuit_copy: Circuit,
+        cycle_and_ops: list[tuple[int, Operation]],
+    ) -> list[Circuit]:
+        '''
+        Given a circuit, create 2^(tree_depth) - 1 circuits that remove up
+        to tree_depth operations. The circuits are sorted by the number of
+        operations removed.
+        '''
```

Comment:
Function and method docstrings should start with a single sentence summary, then more if needed after a blank line, followed by argument descriptions. More sections can follow if necessary, e.g., Returns, Examples, See Also, etc. I know this is just a helper method now, but I can foresee this method being reused in future projects. That is why I am being pedantic here. We should make it as readable as possible.